File: dfs.py
#!/usr/bin/env  python

"""
DFS Try: Was inspired by the AWS interview question video
"""

from collections import defaultdict

def sort_dependencies(dependencies):
    result = []
    visited = set([])
    visiting = set([])

    def dfs(node):
        if node in visited:
            return
        
        visiting.add(node)
        for neighbor in dependencies[node]:
            if neighbor in visiting:
                raise Exception('Cycle Dependency Found')
            if neighbor not in visited:
                dfs(neighbor)
        
        visiting.remove(node)
        visited.add(node)
        result.append(node)

    for node in dependencies.keys():
        dfs(node)

    print(result)

if __name__ == '__main__':
    depDict = {
        0: [],
        1: [0],
        2: [0],
        3: [1,2],
        4: [3],
        5: [1,2],
        6: [5],
        7: [3,4,5,6],
        8: [12],
        9: [0,3],
        10: [2],
        11: [10],
        12: [10,11]
    }
    sort_dependencies(depDict)

File: dfs.cpp
#include <algorithm>
#include <iostream>
#include <map>
#include <list>

using namespace std;

std::list<int>  result;
std::list<int>  visiting;
std::list<int>  visited;

std::map<int,std::list<int>> dependencies = {
                                 { 0, {1} },
                                 { 1, {0} },
                                 { 2, {0} },
                                 { 3, {1,2} },
                                 { 4, {3} },
                                 { 5, {1,2} },
                                 { 6, {5} },
                                 { 7, {3,4,5,6} },
                                 { 8, {12} },
                                 { 9, {0,3} },
                                 { 10, {2} },
                                 { 11, {10} },
                                 { 12, {10,11} }
                              };
void dfs(int node)
{
    if (std::find(visiting.begin(), visiting.end(), node) != visiting.end())    
        return;

    visiting.push_front(node);
    for (int k : dependencies[node])
    {
        if (std::find(visiting.begin(), visiting.end(), k) != visiting.end())
        {
            std::cout << "Cyclic Dependency detected" << std::endl;
            exit(-1);
        }
        if (std::find(visited.begin(), visited.end(), k) == visited.end())
            dfs(k);
    }

    visiting.remove(node);
    visited.push_front(node);

    if (std::find(result.begin(), result.end(), node) == result.end())
        result.push_back(node);
}

void sort_dependencies()
{
    for(std::map<int,std::list<int>>::iterator it = dependencies.begin(); it != dependencies.end(); ++it)
    {
        dfs(it->first);
    }
}

int main(int argc, char **argv)
{
    sort_dependencies();
    for(int n : result)
    {
        std::cout << n << std::endl;
    }
    return 0;
}
